File: epi/SunsetView.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.*;

public class SunsetView {
  public static List<Integer>
  examineBuildingsWithSunset(Iterator<Integer> sequence) {
    //T O(N) worst case still S O(N) but in best case would be S O(1)
    //Min stack
    Deque<int[]> dq = new ArrayDeque<>();
    int i = 0;
    while(sequence.hasNext()) {
      int cur = sequence.next();

      while(!dq.isEmpty() && cur >= dq.peekFirst()[1]) {
        dq.pollFirst();
      }

      dq.offerFirst(new int[]{i++, cur});
    }

    List<Integer> rtList = new ArrayList<>();
    while(!dq.isEmpty()) {
      rtList.add(dq.pollFirst()[0]);
    }
    return rtList;
  }
  @EpiTest(testDataFile = "sunset_view.tsv")
  public static List<Integer>
  examineBuildingsWithSunsetWrapper(List<Integer> sequence) {
    return examineBuildingsWithSunset(sequence.iterator());
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SunsetView.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/SunsetView.kt
package epi.kt

import java.util.*
import kotlin.collections.ArrayList

object SunsetView {
    fun examineBuildingsWithSunset(sequence: Iterator<Int>) : List<Int> {
//        var dq: Deque<Array<Int>> = ArrayDeque()
//        var i = 0
//
//        while(sequence.hasNext()) {
//            val cur = sequence.next()
//            while(!dq.isEmpty() && dq.peekFirst()[1] <= cur) {
//                dq.pollFirst()
//            }
//
//            dq.offerFirst(arrayOf(i++, cur))
//        }
//
//        var rtList: MutableList<Int> = ArrayList()
//        while(!dq.isEmpty()) {
//            rtList.add(dq.pollFirst()[0])
//        }
//
//        return rtList
        var tmp: Deque<Int> = ArrayDeque()
        while(sequence.hasNext()) {
            tmp.offerFirst(sequence.next())
        }
        return examineBuildingsWithSunsetV1(tmp.iterator())
    }

    fun examineBuildingsWithSunsetV1(sequence: Iterator<Int>) : List<Int> {
        var dq = ArrayDeque<Array<Int>>()
        var i = 0

        while(sequence.hasNext()) {
            val cur = sequence.next()
            if(dq.isEmpty() || cur > dq.peekFirst()[1]) {
                dq.offerFirst(arrayOf(i, cur))
            }
            i++
        }

        var rtList: MutableList<Int> = ArrayList()
        while(!dq.isEmpty()) {
            rtList.add((i - 1) - dq.pollLast()[0])
        }

        return rtList
    }
}